File: src/decision_tree.rs
use std::collections::{HashMap, HashSet};

pub enum Node {
    Leaf(String),
    Decision {
        feature_index: usize,
        branches: HashMap<String, Box<Node>>,
    },
}

pub struct DecisionTreeClassifier {
    root: Node,
    default_class: String,  // глобальний клас за замовчуванням (наприклад, найбільш частий у навчанні)
}

impl DecisionTreeClassifier {
    pub fn new() -> Self {
        DecisionTreeClassifier {
            // Спочатку корінь можна тимчасово зробити листком з пустим класом
            root: Node::Leaf(String::new()),
            default_class: String::new(),
        }
    }

    pub fn fit(&mut self, x: &Vec<Vec<String>>, y: &Vec<String>) {
        // Обчислити глобальний переважний клас (для випадку непередбачених значень)
        let mut class_count = HashMap::new();
        for label in y {
            *class_count.entry(label.clone()).or_insert(0) += 1;
        }
        if let Some((majority_class, _)) = class_count.iter().max_by_key(|entry| entry.1) {
            self.default_class = majority_class.clone();
        }

        // Рекурсивно побудувати дерево
        let all_indices: Vec<usize> = (0..y.len()).collect();
        let feature_indices: Vec<usize> = (0..x[0].len()).collect();
        self.root = self.build_tree(x, y, &all_indices, &feature_indices);
    }

    pub fn build_tree(&self, x: &Vec<Vec<String>>, y: &Vec<String>, indices: &[usize], feature_indices: &[usize]) -> Node {
        // 1. Якщо всі приклади одного класу - повертати Leaf
        let first_class = &y[indices[0]];
        let all_same_class = indices.iter().all(|&i| &y[i] == first_class);
        if all_same_class {
            return Node::Leaf(first_class.clone());
        }

        // 2. Якщо не залишилось ознак - повернути Leaf з переважним класом цієї підмножини
        if feature_indices.is_empty() {
            // Знайти найчастіший клас серед indices
            let mut subset_class_count = HashMap::new();
            for &i in indices {
                *subset_class_count.entry(y[i].clone()).or_insert(0) += 1;
            }
            let majority_class = subset_class_count.into_iter().max_by_key(|entry| entry.1).unwrap().0;
            return Node::Leaf(majority_class);
        }

        // Функція для обчислення ентропії списку індексів
        let entropy = |idxs: &[usize]| {
            let total = idxs.len() as f64;
            let mut class_count = HashMap::new();
            for &i in idxs {
                *class_count.entry(y[i].clone()).or_insert(0_usize) += 1;
            }
            let mut ent = 0.0;
            for &count in class_count.values() {
                let p = count as f64 / total;
                if p > 0.0 {
                    ent -= p * p.log2();
                }
            }
            ent
        };

        let base_entropy = entropy(indices);

        // 3. Знайти ознаку з максимальним інформаційним приростом
        let mut best_feature = None;
        let mut best_info_gain = 0.0;
        let mut best_splits: HashMap<String, Vec<usize>> = HashMap::new();
        for &feature in feature_indices {
            // Розбити indices за значеннями ознаки feature
            let mut splits: HashMap<String, Vec<usize>> = HashMap::new();
            for &i in indices {
                let value = &x[i][feature];
                splits.entry(value.clone()).or_insert(Vec::new()).push(i);
            }
            // Обчислити ентропію після розбиття
            let mut new_entropy = 0.0;
            for subset_indices in splits.values() {
                if subset_indices.is_empty() { 
                    continue; 
                }
                let subset_entropy = entropy(subset_indices);
                new_entropy += (subset_indices.len() as f64 / indices.len() as f64) * subset_entropy;
            }
            let info_gain = base_entropy - new_entropy;
            if info_gain > best_info_gain {
                best_info_gain = info_gain;
                best_feature = Some(feature);
                best_splits = splits;
            }
        }

        // Якщо інформаційний приріст нульовий або не знайдено кращої ознаки – листок з переважним класом
        if best_feature.is_none() || best_info_gain <= 0.0 {
            let mut subset_class_count = HashMap::new();
            for &i in indices {
                *subset_class_count.entry(y[i].clone()).or_insert(0) += 1;
            }
            let majority_class = subset_class_count.into_iter().max_by_key(|entry| entry.1).unwrap().0;
            return Node::Leaf(majority_class);
        }

        let best_feature_idx = best_feature.unwrap();
        // 4. Рекурсивно побудувати гілки для кожного значення кращої ознаки
        let mut branches: HashMap<String, Box<Node>> = HashMap::new();
        // Сформувати список доступних ознак для дітей (виключаючи обрану)
        let mut remaining_features: Vec<usize> = feature_indices.to_vec();
        remaining_features.retain(|&f| f != best_feature_idx);
        for (value, subset_indices) in best_splits {
            if subset_indices.is_empty() {
                // Якщо жодного запису з таким значенням в цій підмножині,
                // створити листок з глобальним або локальним переважним класом
                branches.insert(value, Box::new(Node::Leaf(self.default_class.clone())));
            } else {
                // Рекурсивно викликати build_tree для підмножини
                let child_node = self.build_tree(x, y, &subset_indices, &remaining_features);
                branches.insert(value, Box::new(child_node));
            }
        }

        Node::Decision {
            feature_index: best_feature_idx,
            branches,
        }
    }

    pub fn predict(&self, data: &[Vec<String>]) -> Vec<String> {
        let mut predictions = Vec::new();
        for features in data {
            // Прогноз для одного прикладу
            let mut node = &self.root;
            let mut predicted_class = self.default_class.clone();
            loop {
                match node {
                    Node::Leaf(class) => {
                        predicted_class = class.clone();
                        break;
                    }
                    Node::Decision { feature_index, branches } => {
                        let feat_val = if *feature_index < features.len() {
                            &features[*feature_index]
                        } else {
                            // якщо не вистачає ознак у прикладі
                            predicted_class = self.default_class.clone();
                            break;
                        };
                        if let Some(next_node) = branches.get(feat_val) {
                            node = next_node;
                            continue;
                        } else {
                            // Якщо значення відсутнє серед гілок, використовуємо default_class і виходимо
                            predicted_class = self.default_class.clone();
                            break;
                        }
                    }
                }
            }
            predictions.push(predicted_class);
        }
        predictions
    }
}

File: algorithms/decision_tree.py
import numpy as np
from sklearn.base import BaseEstimator, ClassifierMixin

class DecisionTreeClassifierCustom(BaseEstimator, ClassifierMixin):
    def __init__(self, max_depth=None, min_samples_split=2):
        self.tree = None  # дерево буде зберігатись як вкладені словники
        self.max_depth = max_depth  # максимальна глибина дерева
        self.min_samples_split = min_samples_split  # мінімальна кількість прикладів для розбиття
        self.default_class = None  # переважний клас для випадків відсутності значення

    def fit(self, X, y):
        X = np.asarray(X)
        y = np.asarray(y)
        
        # Визначаємо переважний клас у всьому наборі даних
        unique_classes, counts = np.unique(y, return_counts=True)
        self.default_class = unique_classes[np.argmax(counts)]
        
        # Зберігаємо всі класи
        self.classes_ = unique_classes
        
        # Побудувати дерево рекурсивно, починаючи з усіх індексів прикладів і всіх доступних ознак
        features = list(range(X.shape[1]))  # індекси всіх ознак, які ще можна використовувати
        self.tree = self._build_tree(X, y, np.array(range(len(y))), features, depth=0)
        
        return self

    def _build_tree(self, X, y, indices, feature_indices, depth=0):
        # 1. Якщо всі приклади належать до одного класу - повернути листок
        current_classes = y[indices]
        unique_classes = np.unique(current_classes)
        
        if len(unique_classes) == 1:
            return {"label": unique_classes[0].item() if hasattr(unique_classes[0], 'item') else unique_classes[0]}
        
        # 2. Перевірка умов зупинки (max_depth, min_samples_split)
        if (self.max_depth is not None and depth >= self.max_depth) or \
           len(indices) < self.min_samples_split or \
           len(feature_indices) == 0:
            # Повертаємо листок з переважним класом
            classes, counts = np.unique(current_classes, return_counts=True)
            majority_class = classes[np.argmax(counts)]
            return {"label": majority_class.item() if hasattr(majority_class, 'item') else majority_class}

        # 3. Обчислити ентропію поточного вузла
        def entropy(idxs):
            classes_subset = y[idxs]
            _, counts = np.unique(classes_subset, return_counts=True)
            probabilities = counts / len(idxs)
            return -np.sum(probabilities * np.log2(probabilities))

        base_entropy = entropy(indices)

        # 4. Вибрати найкращу ознаку для розбиття (максимальний інформаційний приріст)
        best_feature = None
        best_info_gain = 0.0
        best_splits = None  # збережемо розбиття для найкращої ознаки
        
        for feature in feature_indices:
            # Розбити поточні дані за цією ознакою
            feature_values = X[indices, feature]
            unique_values = np.unique(feature_values)
            
            # Словник для зберігання розбиття
            splits = {}
            
            # Для кожного значення ознаки знаходимо відповідні індекси
            for value in unique_values:
                # Конвертуємо значення в хешований тип
                hashable_value = value.item() if hasattr(value, 'item') else value
                
                # Знаходимо індекси, де значення ознаки дорівнює поточному
                # Спочатку знаходимо маску для значень
                mask = feature_values == value
                # Потім застосовуємо маску до індексів
                value_indices = indices[mask]
                
                splits[hashable_value] = value_indices
            
            # Обчислити ентропію після розбиття (зважену суму ентропій підмножин)
            new_entropy = 0.0
            for subset_indices in splits.values():
                if len(subset_indices) == 0:
                    continue
                subset_entropy = entropy(subset_indices)
                new_entropy += (len(subset_indices) / len(indices)) * subset_entropy
                
            info_gain = base_entropy - new_entropy
            if info_gain > best_info_gain:
                best_info_gain = info_gain
                best_feature = feature
                best_splits = splits

        if best_feature is None or best_info_gain <= 0:
            # Якщо не вдалося знайти розбиття, повертаємо листок з переважним класом
            classes, counts = np.unique(current_classes, return_counts=True)
            majority_class = classes[np.argmax(counts)]
            return {"label": majority_class.item() if hasattr(majority_class, 'item') else majority_class}

        # 5. Створити піддерева для кожного значення найкращої ознаки
        tree = {
            "feature": best_feature, 
            "branches": {},
            "majority_class": self.default_class.item() if hasattr(self.default_class, 'item') else self.default_class
        }
        
        # Зберігаємо переважний клас в цьому вузлі для випадків, коли зустрінеться нове значення
        classes, counts = np.unique(current_classes, return_counts=True)
        tree["majority_class"] = classes[np.argmax(counts)].item() if hasattr(classes[np.argmax(counts)], 'item') else classes[np.argmax(counts)]
        
        # Видалити використану ознаку зі списку доступних
        remaining_features = [f for f in feature_indices if f != best_feature]
        
        for value, subset_indices in best_splits.items():
            if len(subset_indices) == 0:
                # Якщо підмножина порожня, привласнити листок з переважним класом поточного вузла
                tree["branches"][value] = {"label": tree["majority_class"]}
            else:
                # Рекурсивно побудувати піддерево для підмножини
                tree["branches"][value] = self._build_tree(X, y, subset_indices, remaining_features, depth + 1)
                
        return tree

    def predict(self, X):
        X = np.asarray(X)
        
        if len(X.shape) == 1:
            # Один приклад
            return np.array([self._predict_single(X)])
        else:
            # Набір прикладів
            return np.array([self._predict_single(features) for features in X])

    def _predict_single(self, features):
        if self.tree is None:
            return self.default_class
            
        node = self.tree
        while "label" not in node:
            feature_idx = node["feature"]
            value = features[feature_idx]
            
            hashable_value = value.item() if hasattr(value, 'item') else value
            
            if hashable_value not in node["branches"]:
                return node["majority_class"]
                
            node = node["branches"][hashable_value]
            
        return node["label"]
    
   